fastlane_bot/tools/optimizer/margpoptimizer.py

```diff
@@ -155,247 +187,288 @@ def optimize(self, sfc=None, result=None, *, params=None):
+                if P("debug_dtknd"):
                     dtkn_d, dtkn = dtknfromp_f(plog10, islog10=True, asdct=True)
                 else:
                     dtkn = dtknfromp_f(plog10, islog10=True, asdct=False)
 
                 # calculate the Jacobian
-                # if P("debug"):
-                #     print("\n[margp_optimizer] ============= JACOBIAN =============>>>")
-                J = self.J(dtknfromp_f, plog10)  
-                    # ATTENTION: dtknfromp_f takes log10(p) as input
-                if P("debug"):
-                    # print("==== J ====>")
-                    print("\n============= JACOBIAN =============>>>")
+                
+                J = self.J(dtknfromp_f, plog10, jach=jach)  
+                    # ATTENTION: dtknfromp_f takes log10(p) as input by default
+                
+                if P("debug_j"):
+                    print("\n[margp_optimizer]\n============= JACOBIAN =============>>>")
                     print(J)
-                    # print("<=== J =====")
                     print("<<<============= JACOBIAN =============\n")
                 
                 # Update p, dtkn using the Newton-Raphson formula
                 try:
                     dplog10 = np.linalg.solve(J, -dtkn)
+                
                 except np.linalg.LinAlgError:
+                    
                     if P("verbose") or P("debug"):
-                        print("[margp_optimizer] singular Jacobian, using lstsq instead")
+                        print("\n[margp_optimizer] singular Jacobian, using lstsq instead")
+                    
                     dplog10 = np.linalg.lstsq(J, -dtkn, rcond=None)[0]
-                    # https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html
-                    # https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html
+                        # https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html
+                        # https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html
+                
+                # #### TODO: EXPERIMENTAL: ADD A DAMPING FACTOR TO THE JACOBIAN
                 
-                # update log prices, prices and determine the criterium...
-                p0log10 = [*plog10]
-                plog10 += dplog10
-                p = np.exp(plog10 * np.log(10))
-                criterium = np.linalg.norm(dplog10)
+                # #dplog10 = np.clip(dplog10, -0.1, 0.1)
+                # nrm = normf(dplog10)
+                # if nrm > 0.1:
+                #     dplog10 /= nrm
+                
+                # #### END EXPERIMENTAL
+                
+                # update log prices, prices...
+                p0log10 = [*plog10]                 # keep current log prices (deep copy)
+                plog10 += dplog10                   # update log prices
+                p = np.exp(plog10 * np.log(10))     # expand log to actual prices
```

Comment:
If this piece of code is used excessively, then calculating the value of `np.log(10)` in advance and using it as a constant may serve for better performance.

Reply:
yeah; was thinking about this as well; I tend to avoid numeric optimization on a gut feel basis those days: there is not need to make code harder to read and more error prone for a 0.01% improvement in performance. I may have mentioned a few (hundred) times that I really really would love to profile that code in a live context